File: api-gin/internal/routers/router.go
package routers

import (
	v1 "admin-api/internal/api/v1"
	"admin-api/internal/middleware"

	"github.com/gin-gonic/gin"
)

func Init(r *gin.Engine) {
	r.Use(middleware.Recovery)
	r.Use(middleware.Logger)
	r.Use(middleware.Access)

	r.POST("login", v1.Auth.Login)
	r.POST("renewToken", v1.Auth.RenewToken)

	api := r.Group("")
	api.Use(middleware.Auth)
	{
		authApi := api.Group("auth")
		{
			authApi.GET("jwt", v1.Auth.JwtInfo)
			authApi.GET("menu", v1.Auth.MenuTree)
			authApi.POST("password", v1.Auth.ChangePwd)
		}

		groupApi := api.Group("group")
		{
			groupApi.GET("", v1.Auth.Group)
			groupApi.POST("", v1.Auth.AddGroup)
			groupApi.POST("edit", v1.Auth.ChangeGroup)
			groupApi.POST("user", v1.Auth.SetGroupUser)
			groupApi.POST("menu", v1.Auth.SetGroupMenu)
		}

		menuApi := api.Group("menu")
		{
			menuApi.GET("", v1.Auth.Menu)
			menuApi.GET("info", v1.Auth.MenuInfo)
			menuApi.POST("group", v1.Auth.SetMenuGroup)
		}

		userApi := api.Group("user")
		{
			userApi.GET("self", v1.User.Self)
			userApi.GET("list", v1.User.List)
			userApi.GET("info", v1.User.Info)
			userApi.POST("self", v1.User.ChangeSelf)
			userApi.POST("add", v1.User.Add)
			userApi.POST("update", v1.User.Change)
			userApi.POST("status", v1.User.ChangeStatus)
			userApi.POST("group", v1.User.SetUserGroup)
		}
	}

}

File: api-ts/src/controllers/router.ts
import Router from 'koa-router'
import Auth from './Auth'
import User from './User'

const routerMap: RouterMap = {
  '': [
    { method: 'post', path: '/login', ctl: Auth.login },
    { method: 'post', path: '/renewToken', ctl: Auth.renewToken },
  ],

  'auth': [
    { method: 'get', path: '/jwt', ctl: Auth.jwtInfo },
    { method: 'get', path: '/menu', ctl: Auth.menu },
    { method: 'post', path: '/password', ctl: Auth.changPwd },
  ],

  'group': [
    { method: 'get', path: '/', ctl: Auth.groupList },
    { method: 'post', path: '/', ctl: Auth.addGroup },
    { method: 'post', path: '/edit', ctl: Auth.editGroup },
    { method: 'post', path: '/user', ctl: Auth.setGroupUser },
    { method: 'post', path: '/menu', ctl: Auth.setGroupMenu },
  ],

  'menu': [
    { method: 'get', path: '/', ctl: Auth.menuTree },
    { method: 'get', path: '/info', ctl: Auth.menuInfo },
    { method: 'post', path: '/group', ctl: Auth.setMenuGroup },
  ],

  'user': [
    { method: 'get', path: '/self', ctl: User.self },
    { method: 'post', path: '/self', ctl: User.editSelf },
    { method: 'get', path: '/list', ctl: User.list },
    { method: 'get', path: '/info', ctl: User.info },
    { method: 'post', path: '/add', ctl: User.add },
    { method: 'post', path: '/update', ctl: User.updateInfo },
    { method: 'post', path: '/status', ctl: User.status },
    { method: 'post', path: '/group', ctl: User.setGroup },
  ]

}

const methodMap: { [key in RouteConf['method']]: (router: Router, path: string, ctl: Router.IMiddleware) => Router } = {
  get: (router, path, ctl) => router.get(path, ctl),
  post: (router, path, ctl) => router.post(path, ctl),
  put: (router, path, ctl) => router.put(path, ctl),
  del: (router, path, ctl) => router.del(path, ctl),
  all: (router, path, ctl) => router.all(path, ctl),
}

const routers = new Router()

for (const k in routerMap) {
  const routes = routerMap[k]
  const r = new Router()
  routes.forEach(route => registerRoute(r, route))
  routers.use(k == '' ? '' : `/${k}`, r.routes(), r.allowedMethods())
}

export default routers

function registerRoute(router: Router, route: RouteConf) {
  const { method, path, ctl } = route
  if (methodMap[method]) {
    methodMap[method](router, path, ctl)
  } else {
    throw new Error(`Unsupported method: ${method}`)
  }
}
